src/plugins/mission_raw/mission_raw_impl.cpp

```diff
@@ -156,63 +249,217 @@ MissionRawImpl::convert_items(const std::vector<MAVLinkMissionTransfer::ItemInt>
+            });
+        });
+}
+
+void MissionRawImpl::set_current_mission_item_async(
+    int current_mavlink, MissionRaw::result_callback_t& callback)
+{
+    if (current_mavlink < 0 && current_mavlink > _mission_data.last_total_mavlink_mission_item) {
         _parent->call_user_callback([callback]() {
             if (callback) {
-                callback(MissionRaw::Result::UNSUPPORTED);
+                callback(MissionRaw::Result::INVALID_ARGUMENT);
+                return;
             }
         });
-        return;
     }
 
-    const auto int_items = convert_to_int_items(mission_raw);
-
-    _parent->mission_transfer().upload_items_async(
-        MAV_MISSION_TYPE_MISSION,
-        int_items,
-        [this, callback, int_items](MAVLinkMissionTransfer::Result result) {
+    _parent->mission_transfer().set_current_item_async(
+        current_mavlink, [this, callback](MAVLinkMissionTransfer::Result result) {
             auto converted_result = convert_result(result);
-            auto converted_items = convert_items(int_items);
-            _parent->call_user_callback([callback, converted_result, converted_items]() {
+            _parent->call_user_callback([callback, converted_result]() {
                 if (callback) {
                     callback(converted_result);
                 }
             });
         });
 }
 
+void MissionRawImpl::report_progress()
+{
+    const auto temp_callback = _mission_data.progress_callback;
+    if (temp_callback == nullptr) {
+        return;
+    }
+
+    int current = current_mavlink_mission_item();
+    int total = total_mavlink_mission_items();
+
+    bool should_report = false;
+    {
+        std::lock_guard<std::mutex> lock(_mission_data.mutex);
+        if (_mission_data.last_current_reported_mavlink_mission_item != current) {
+            _mission_data.last_current_reported_mavlink_mission_item = current;
+            should_report = true;
+        }
+        if (_mission_data.last_total_reported_mavlink_mission_item != total) {
+            _mission_data.last_total_reported_mavlink_mission_item = total;
+            should_report = true;
+        }
+    }
+
+    if (should_report) {
+        std::lock_guard<std::mutex> lock(_mission_data.mutex);
+        _parent->call_user_callback([temp_callback, current, total]() {
+            LogDebug() << "current: " << current << ", total: " << total;
+            temp_callback(current, total);
+        });
+    }
+}
+
+bool MissionRawImpl::is_mission_finished() const
+{
+    std::lock_guard<std::mutex> lock(_mission_data.mutex);
+
+    if (_mission_data.last_current_mavlink_mission_item < 0) {
+        return false;
+    }
+
+    if (_mission_data.last_reached_mavlink_mission_item < 0) {
+        return false;
+    }
+
+    if (_mission_data.last_total_mavlink_mission_item < 0) {
+        return false;
+    }
+
+    return (
+        _mission_data.last_reached_mavlink_mission_item ==
+        _mission_data.last_total_mavlink_mission_item);
+}
```

Comment:
@julianoes Not sure how to manage the `mission_finished` and if we should manage it in the `MissionRaw`. Especially with the RTL, in mission_impl there is something like this:

```
    // It is not straightforward to look at "current" because it jumps to 0
    // once the last item has been done. Therefore we have to lo decide using
    // "reached" here.
    // It seems that we never receive a reached when RTL is initiated after
    // a mission, and we need to account for that.
    const unsigned rtl_correction = _enable_return_to_launch_after_mission ? 2 : 1;

    return (
        unsigned(_mission_data.last_reached_mavlink_mission_item + rtl_correction) ==
        _mission_data.mavlink_mission_item_to_mission_item_indices.size());
```

but I guess that not something we can do in `MissionRaw`. So we should rather return `reached_mavlink_mission_item` as well and let the user manage when the mission is finished ?

Note: it should be ```
_mission_data.last_reached_mavlink_mission_item + 1 ==
        _mission_data.last_total_mavlink_mission_item);
``` but I'll update it when I know what to do with the `mission_finished` function

Reply:
I'd have to play with SITL and see how it works, I'm not sure about the edge cases from the top of my head, sorry.

Reply:
I have something like this if I upload just 6 waypoints (without changing speed or anything and without RTL)

Start Mission
[04:13:45|Debug] mavlink_current: 1, mavlink_total: 6 (mission_raw_impl.cpp:370)
[04:13:45|Debug] mavlink_reached: 0, mavlink_total: 6 (mission_raw_impl.cpp:402)
[04:13:45|Debug] 1st waypoint reached
[04:13:50|Debug] mavlink_reached: 1, mavlink_total: 6 (mission_raw_impl.cpp:402)
[04:13:50|Debug] mavlink_current: 2, mavlink_total: 6 (mission_raw_impl.cpp:370)
[04:13:50|Debug] 2nd waypoint reached
[04:13:55|Debug] mavlink_reached: 2, mavlink_total: 6 (mission_raw_impl.cpp:402)
[04:13:55|Debug] mavlink_current: 3, mavlink_total: 6 (mission_raw_impl.cpp:370)
[04:13:55|Debug] 3rd waypoint reached
[04:14:01|Debug] mavlink_reached: 3, mavlink_total: 6 (mission_raw_impl.cpp:402)
[04:14:01|Debug] mavlink_current: 4, mavlink_total: 6 (mission_raw_impl.cpp:370)
[04:14:01|Debug] 4th waypoint reached
[04:14:06|Debug] mavlink_reached: 4, mavlink_total: 6 (mission_raw_impl.cpp:402)
[04:14:06|Debug] mavlink_current: 5, mavlink_total: 6 (mission_raw_impl.cpp:370)
[04:14:06|Debug] 5th waypoint reached
[04:14:11|Debug] MAVLink: info: Mission finished, loitering. (system_impl.cpp:257)
[04:14:11|Debug] mavlink_reached: 5, mavlink_total: 6 (mission_raw_impl.cpp:402)
[04:14:11|Debug] 6th waypoint reached (and last)

Ok let me know ! Not sure if the `mission_finished` would take all the possibilities that an user could generate...

Reply:
and with RTL:
Start Mission
[05:14:20|Debug] mavlink_reached: 0, mavlink_total: 7 (mission_raw_impl.cpp:402)
[05:14:20|Debug] mavlink_current: 1, mavlink_total: 7 (mission_raw_impl.cpp:370)
[05:14:20|Debug] 1st waypoint reached
[05:14:24|Debug] mavlink_reached: 1, mavlink_total: 7 (mission_raw_impl.cpp:402)
[05:14:24|Debug] mavlink_current: 2, mavlink_total: 7 (mission_raw_impl.cpp:370)
[05:14:24|Debug] 2nd waypoint reached
[05:14:30|Debug] mavlink_reached: 2, mavlink_total: 7 (mission_raw_impl.cpp:402)
[05:14:30|Debug] mavlink_current: 3, mavlink_total: 7 (mission_raw_impl.cpp:370)
[05:14:30|Debug] 3rd waypoint reached
[05:14:35|Debug] mavlink_reached: 3, mavlink_total: 7 (mission_raw_impl.cpp:402)
[05:14:35|Debug] mavlink_current: 4, mavlink_total: 7 (mission_raw_impl.cpp:370)
[05:14:35|Debug] 4th waypoint reached
[05:14:40|Debug] mavlink_reached: 4, mavlink_total: 7 (mission_raw_impl.cpp:402)
[05:14:40|Debug] mavlink_current: 5, mavlink_total: 7 (mission_raw_impl.cpp:370)
[05:14:40|Debug] 5th waypoint reached
[05:14:45|Debug] MAVLink: info: Returning to launch (system_impl.cpp:257)
[05:14:45|Debug] mavlink_reached: 5, mavlink_total: 7 (mission_raw_impl.cpp:402)
[05:14:45|Debug] mavlink_current: 6, mavlink_total: 7 (mission_raw_impl.cpp:370)
[05:14:45|Debug] 6th waypoint reached (and last)
[05:14:45|Debug] MAVLink: info: RTL HOME activated (system_impl.cpp:257)
[05:14:45|Debug] MAVLink: info: RTL: climb to 518 m (31 m above home) (system_impl.cpp:257)
Flight mode: RTL

Reply:
So you're saying it's all working? :smile: Or what am I reading?

Reply:
The progress current and reached are working fine but I did not implement the `mission_finished`. The "waypoint reached" are generated by my companion computer and not `MissionRaw` so `MissionRaw` only gives me updates on reached/current

Reply:
So I did not implement the function `is_mission_finished` below from `mission_impl` because to keep the Mission**Raw** I'm not sure if it makes sense to detect if there is a RTL command in mission uploaded.

```
bool MissionImpl::is_mission_finished() const
{
    std::lock_guard<std::recursive_mutex> lock(_mission_data.mutex);

    if (_mission_data.last_current_mavlink_mission_item < 0) {
        return false;
    }

    if (_mission_data.last_reached_mavlink_mission_item < 0) {
        return false;
    }

    if (_mission_data.mavlink_mission_item_to_mission_item_indices.size() == 0) {
        return false;
    }

    // It is not straightforward to look at "current" because it jumps to 0
    // once the last item has been done. Therefore we have to lo decide using
    // "reached" here.
    // It seems that we never receive a reached when RTL is initiated after
    // a mission, and we need to account for that.
    const unsigned rtl_correction = _enable_return_to_launch_after_mission ? 2 : 1;

    return (
        unsigned(_mission_data.last_reached_mavlink_mission_item + rtl_correction) ==
        _mission_data.mavlink_mission_item_to_mission_item_indices.size());
}
```

Which mean I do not have `current_mission_item` from `Mission` but I replaced it by `current_mavlink_mission_item` and `reached_mavlink_mission_item` to let the developer using `MissionRaw` in a "raw" way

Reply:
So for me it's all good ! Now this the test part that I never have the time to do 😅